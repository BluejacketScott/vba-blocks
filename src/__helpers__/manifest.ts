import { dir as FIXTURES } from '../../tests/__fixtures__';
import { Manifest } from '../manifest';
import { isPathDependency } from '../manifest/dependency';
import { relative } from '../utils/path';

export function normalizeManifest(manifest: Manifest, relativeTo: string = FIXTURES): Manifest {
  for (const src of manifest.src) {
    src.path = normalizePath(src.path, relativeTo);

    if (src.binary) {
      src.binary = normalizePath(src.binary, relativeTo);
    }
  }
  for (const dependency of manifest.dependencies) {
    if (isPathDependency(dependency)) {
      dependency.path = normalizePath(dependency.path, relativeTo);
    }
  }
  for (const src of manifest.devSrc) {
    src.path = normalizePath(src.path, relativeTo);

    if (src.binary) {
      src.binary = normalizePath(src.binary, relativeTo);
    }
  }
  for (const dependency of manifest.devDependencies) {
    if (isPathDependency(dependency)) {
      dependency.path = normalizePath(dependency.path, relativeTo);
    }
  }

  if (manifest.target) {
    manifest.target.path = normalizePath(manifest.target.path, relativeTo);
  }

  return manifest;
}

function normalizePath(path: string, relativeTo: string = FIXTURES): string {
  return relative(relativeTo, path);
}
