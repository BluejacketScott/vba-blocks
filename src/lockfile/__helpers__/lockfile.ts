export function normalizeLockfile(lockfile: string): string {
	return lockfile.replace(/vba\-blocks v\d.\d.\d/g, "vba-blocks v#.#.#");
}
