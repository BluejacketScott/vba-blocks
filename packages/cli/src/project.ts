import { join } from 'path';
import env from './env';
import { Config, loadConfig } from './config';
import { Manifest, Source, Reference, loadManifest } from './manifest';
import { Workspace } from './workspace';
import SourceManager, { Registration } from './sources';
import resolve, { DependencyGraph, getRegistration } from './resolve';

import { parallel } from './utils';

export interface Project {
  manifest: Manifest;
  workspace: Workspace;
  packages: DependencyGraph;

  config: Config;
  paths: {
    root: string;
    dir: string;
    build: string;
    backup: string;
  };
}

export async function loadProject(dir?: string): Promise<Project> {
  const manifest = await loadManifest(dir || env.cwd);

  // TODO Load workspace
  const workspace: Workspace = {
    root: manifest,
    members: []
  };

  const config = await loadConfig(manifest, workspace);
  const packages = await resolve(config, workspace);

  const paths = {
    root: workspace.root.dir,
    dir: manifest.dir,
    build: join(manifest.dir, 'build'),
    backup: join(manifest.dir, 'build', '.backup')
  };

  return {
    manifest,
    workspace,
    packages,
    config,
    paths
  };
}

export async function fetchDependencies(project: Project): Promise<Manifest[]> {
  const manager = new SourceManager(project.config);
  const manifests = await parallel(
    project.manifest.dependencies,
    async dependency => {
      const registration = getRegistration(project.packages, dependency);
      const path = await manager.fetch(registration);
      const manifest = await loadManifest(path);

      return manifest;
    }
  );

  return manifests;
}
