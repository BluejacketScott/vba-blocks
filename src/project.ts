import env from './env';
import { join, normalize } from './utils/path';
import { tmpFolder } from './utils/fs';
import parallel from './utils/parallel';
import { Config, loadConfig } from './config';
import { Manifest, loadManifest } from './manifest';
import { Workspace, loadWorkspace } from './workspace';
import { fetch } from './sources';
import resolve, { DependencyGraph } from './resolve';
import { readLockfile, isLockfileValid } from './lockfile';

const debug = require('debug')('vba-blocks:project');

export interface Project {
  manifest: Manifest;
  workspace: Workspace;
  packages: DependencyGraph;

  config: Config;
  paths: {
    root: string;
    dir: string;
    build: string;
    backup: string;
    staging: string;
  };
  has_dirty_lockfile: boolean;
}

/**
 * Load project from given directory / cwd
 *
 * 1. Load manifest in dir
 * 2. Load workspace from manifest
 * 3. Load lockfile
 * 4. Resolve dependencies
 */
export async function loadProject(dir: string = env.cwd): Promise<Project> {
  const manifest = await loadManifest(dir);

  const config = await loadConfig();
  const workspace = await loadWorkspace(manifest);
  const lockfile = await readLockfile(workspace.root.dir);

  // Resolve packages from lockfile or from sources
  const has_dirty_lockfile = !lockfile || !isLockfileValid(lockfile, workspace);
  debug(
    !has_dirty_lockfile
      ? 'Loading packages from lockfile'
      : 'Resolving packages'
  );

  const packages = !has_dirty_lockfile
    ? lockfile!.packages
    : await resolve(config, workspace, lockfile ? lockfile.packages : []);

  const paths = {
    root: workspace.root.dir,
    dir: manifest.dir,
    build: join(manifest.dir, 'build'),
    backup: join(manifest.dir, 'build', '.backup'),
    staging: await tmpFolder({ dir: env.staging })
  };

  return {
    manifest,
    workspace,
    packages,
    config,
    paths,
    has_dirty_lockfile
  };
}

export interface FetchProject {
  manifest: Manifest;
  packages: DependencyGraph;
  config: Config;
}

/**
 * Fetch all dependencies for project
 * (based on already resolved project.packages)
 *
 * After sources complete fetches, manifests are loaded and returned for each package
 */
export async function fetchDependencies(
  project: FetchProject
): Promise<Manifest[]> {
  const manifests = await parallel(
    project.packages,
    async registration => {
      const path = await fetch(project.config.sources, registration);
      const manifest = await loadManifest(path);

      return manifest;
    },
    { progress: env.reporter.progress('Fetch dependencies') }
  );

  return manifests;
}

export interface ProjectOptions {
  type?: 'project' | 'package';
}

/**
 * Initialize new project
 *
 * Minimum requirements: name and dir
 */
export async function initProject(
  name: string,
  dir: string,
  options: ProjectOptions = {}
): Promise<Project> {
  dir = normalize(dir);
  const { type = 'project' } = options;

  const config = await loadConfig();

  // TODO load defaults from config
  const version = '0.1.0';
  const authors: string[] = [];
  const license = 'MIT';

  // Manually generate manifest and project
  // (may be included in project in the future)
  const manifest: Manifest = {
    type,
    name,
    version,
    metadata: {
      authors,
      publish: type === 'package',
      license
    },
    dependencies: [],
    src: [],
    references: [],
    targets: [],
    dir
  };

  const workspace = await loadWorkspace(manifest);

  const project: Project = {
    manifest,
    workspace,
    packages: [],
    config,
    paths: {
      root: workspace.root.dir,
      dir,
      build: join(dir, 'build'),
      backup: join(dir, 'build', '.backup'),
      staging: await tmpFolder({ dir: env.staging })
    },
    has_dirty_lockfile: true
  };

  return project;
}
