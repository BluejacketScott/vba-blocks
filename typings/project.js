import env from './env';
import { join, normalize } from './utils/path';
import { tmpFolder } from './utils/fs';
import parallel from './utils/parallel';
import { loadConfig } from './config';
import { loadManifest } from './manifest';
import { loadWorkspace } from './workspace';
import { fetch } from './sources';
import resolve from './resolve';
import { readLockfile, isLockfileValid } from './lockfile';
import { fetchingDependencies } from './messages';
const debug = require('debug')('vba-blocks:project');
/**
 * Load project from given directory / cwd
 *
 * 1. Load manifest in dir
 * 2. Load workspace from manifest
 * 3. Load lockfile
 * 4. Resolve dependencies
 */
export async function loadProject(dir = env.cwd) {
    const manifest = await loadManifest(dir);
    const config = await loadConfig();
    const workspace = await loadWorkspace(manifest);
    const lockfile = await readLockfile(workspace.root.dir);
    // Resolve packages from lockfile or from sources
    const has_dirty_lockfile = !lockfile || !(await isLockfileValid(lockfile, workspace));
    debug(!has_dirty_lockfile ? 'Loading packages from lockfile' : 'Resolving packages');
    const packages = !has_dirty_lockfile
        ? lockfile.packages
        : await resolve(config, workspace, lockfile ? lockfile.packages : []);
    const paths = {
        root: workspace.root.dir,
        dir: manifest.dir,
        build: join(manifest.dir, 'build'),
        backup: join(manifest.dir, 'build', '.backup'),
        staging: await tmpFolder({ dir: env.staging })
    };
    return {
        manifest,
        workspace,
        packages,
        config,
        paths,
        has_dirty_lockfile
    };
}
/**
 * Fetch all dependencies for project
 * (based on already resolved project.packages)
 *
 * After sources complete fetches, manifests are loaded and returned for each package
 */
export async function fetchDependencies(project) {
    const manifests = await parallel(project.packages, async (registration) => {
        const path = await fetch(project.config.sources, registration);
        const manifest = await loadManifest(path);
        return manifest;
    }, { progress: env.reporter.progress(fetchingDependencies()) });
    return manifests;
}
/**
 * Initialize new project
 *
 * Minimum requirements: name and dir
 */
export async function initProject(name, dir, options = {}) {
    dir = normalize(dir);
    const { type = 'project' } = options;
    const config = await loadConfig();
    // TODO load defaults from config
    const version = '0.0.0';
    const authors = [];
    const license = 'UNLICENSED';
    const __temp_defaults = type === 'package' ? [] : ['publish'];
    // Manually generate manifest and project
    // (may be included in project in the future)
    const manifest = {
        type,
        name,
        version,
        metadata: {
            authors,
            publish: type === 'package',
            license,
            __temp_defaults
        },
        dependencies: [],
        src: [],
        references: [],
        dir
    };
    const workspace = await loadWorkspace(manifest);
    const project = {
        manifest,
        workspace,
        packages: [],
        config,
        paths: {
            root: workspace.root.dir,
            dir,
            build: join(dir, 'build'),
            backup: join(dir, 'build', '.backup'),
            staging: await tmpFolder({ dir: env.staging })
        },
        has_dirty_lockfile: true
    };
    return project;
}
