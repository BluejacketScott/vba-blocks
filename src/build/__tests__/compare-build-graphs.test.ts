import { setup, reset } from '../../../tests/__helpers__/project';
import { dir, complex, standardExport, standardChangesExport } from '../../../tests/__fixtures__';
import { normalizeComponent } from '../component';
import loadFromProject from '../load-from-project';
import loadFromExport from '../load-from-export';
import compareBuildGraphs, { Changeset } from '../compare-build-graphs';

afterAll(reset);

test('should find no changes between build graphs', async () => {
  const { project, dependencies } = await setup(complex);

  const before = await loadFromProject(project, dependencies);
  const after = await loadFromExport(standardExport);

  const changeset = compareBuildGraphs(before, after);

  expect(changeset.components.added.length).toEqual(0);
  expect(changeset.components.changed.length).toEqual(0);
  expect(changeset.components.removed.length).toEqual(0);
});

test('should find added, changed, and removed between build graphs', async () => {
  const { project, dependencies } = await setup(complex);

  const before = await loadFromProject(project, dependencies);
  const after = await loadFromExport(standardChangesExport);

  const changeset = compareBuildGraphs(before, after);
  expect(normalizeChangeset(changeset)).toMatchSnapshot();

  // TODO finds UserForm1 as changed, look into why
});

export function normalizeChangeset(changeset: Changeset): Changeset {
  const { components, references } = changeset;

  const added = components.added.map(component => normalizeComponent(component, dir));
  const changed = components.changed.map(component => normalizeComponent(component, dir));
  const removed = components.removed.map(component => normalizeComponent(component, dir));

  return {
    components: { added, changed, removed },
    references
  };
}
