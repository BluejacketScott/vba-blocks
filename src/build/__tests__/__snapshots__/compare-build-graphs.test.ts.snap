// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should find added, changed, and removed between build graphs 1`] = `
Object {
  "components": Object {
    "added": Array [
      Object {
        "binary_path": undefined,
        "code": "Attribute VB_Name = \\"Added\\"
' (added)
",
        "details": Object {
          "dependency": undefined,
          "path": undefined,
        },
        "filename": "Added.bas",
        "name": "Added",
        "type": "module",
      },
    ],
    "changed": Array [
      Object {
        "binary_path": undefined,
        "code": "VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = \\"ThisWorkbook\\"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = True
Attr...",
        "details": Object {
          "dependency": undefined,
          "path": "projects/complex/src/ThisWorkbook.cls",
        },
        "filename": "ThisWorkbook.cls",
        "name": "ThisWorkbook",
        "type": "class",
      },
    ],
    "removed": Array [
      Object {
        "binary_path": undefined,
        "code": "Attribute VB_Name = \\"Validation\\"
Public Function Validate(Args As Variant) As String
    On Error GoTo ErrorHandling

    Dim Errors As New Collection

    If ThisWorkbook.VBProject.VBComponents...",
        "details": Object {
          "dependency": undefined,
          "path": "projects/complex/src/Validation.bas",
        },
        "filename": "Validation.bas",
        "name": "Validation",
        "type": "module",
      },
    ],
  },
  "references": Object {
    "added": Array [],
    "changed": Array [],
    "removed": Array [],
  },
}
`;
